### PR TITLE
Add script to compute number of ABOi HLAi pairs (report pro Xenii).

### PR DESCRIPTION
**Pull request by jakubmonhart**

### CODE REVIEW

**Comment by jakubmonhart** [comment]
> 3. Could you specify which criteria are used to identify which
> exchanges will proceed for the transplantation? How do you apply
> these criteria, e.g. is it sequential, or are the weights provided
> for each transplants based on some characteristics (if so, which)?

We first check sequentially if the transplant is possible. This is done based on several criteria:
- Donor must have blood group that is compatible with the recipient, or a blood group that is acceptable for the recipient (acceptable blood groups are sometimes specified as the transplant can proceed even in case of ABOi)
- There must be no positive crossmatch of donor's hla antigens and recipient's hla antibodies.
- Exchanges between some countries are not possible.

If transplant is possible, we compute compatibility index of the donor and recipient based on how well do their HLA typings match. We sometimes further check if this compatibility index is higher then the compatibility index between the donor that is related to the recipient and do not consider the exchange otherwise.

Remaining possible exchanges are scored based on the compatibility indices (+ some bonus for blood group compatibility) and an optimal matching is found.

**Comment by jakubmonhart** [comment]
Tady mam nejaky sample output toho co nam ten vpra pocitac dava:

```
recipient.medical_id='CAN_0R'
unacceptable_antibodies=['A*02:04', 'A*02:09', 'A*02:20', 'A*11:09', 'A*24:20', 'A*25:01', 'A*66:03', 'A*68:08', 'B*14:06', 'B*15:146', 'B*15:20', 'B*15:46', 'B*35:44', 'B*39:01', 'B*44:06', 'B*48:04', 'B*48:07', 'B*48:10', 'B*58:20', 'DRB1*01:01', 'DRB1*03:01', 'DRB1*03:40', 'DRB1*04:23', 'DRB1*04:26', 'DRB1*04:32', 'DRB1*11:14', 'DRB1*11:37', 'DRB1*13:66', 'DRB1*14:14', 'DRB1*14:19', 'DPB1*92:01', 'DQA1*02:01', 'DQB1*02:02', 'DQB1*06:12']
/usr/local/anaconda3/envs/txmatching/lib/python3.10/site-packages/bs4/builder/__init__.py:545: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  warnings.warn(
<frequency>58.94</frequency>
<matches>5894</matches>
<panelsize>10000</panelsize>
<errormessage>Warning: antigen(s) invalid and not used in the calculation: A*02:09 A*11:09 A*24:20 A*66:03 A*68:08 B*14:06 B*15:146 B*15:46 B*35:44 B*48:04 B*48:07 B*48:10 B*58:20 DRB1*03:40 DRB1*04:23 DRB1*04:26 DRB1*04:32 DRB1*11:37 DRB1*13:66 DRB1*14:14 DRB1*14:19 DQB1*06:12</errormessage>

recipient.medical_id='CAN_2R'
unacceptable_antibodies=['A*01:07', 'A*02:77', 'A*03:24', 'A*24:05', 'A*24:13', 'A*30:15', 'A*33:01', 'A*68:08', 'B*07:10', 'B*14:06', 'B*15:146', 'B*15:19', 'B*15:20', 'B*15:46', 'B*18:119', 'B*39:01', 'B*58:01', 'B*58:20', 'DRB1*04:32', 'DRB1*08:16', 'DRB1*13:01', 'DRB1*14:14', 'DRB1*14:20', 'DRB1*14:49', 'DPA1*02:07', 'DPA1*02:18', 'DPB1*23:01', 'DPB1*59:01', 'DPB1*80:01', 'DPB1*89:01', 'DPB1*92:01', 'DPB1*94:01', 'DQA1*01:25', 'DQA1*04:01', 'DQA1*04:05', 'DQA1*05:03', 'DQB1*03:02', 'DQB1*06:03']
<frequency>41.53</frequency>
<matches>4153</matches>
<panelsize>10000</panelsize>
<errormessage>Warning: antigen(s) invalid and not used in the calculation: A*01:07 A*02:77 A*03:24 A*24:05 A*24:13 A*30:15 A*68:08 B*14:06 B*15:146 B*15:19 B*15:46 B*18:119 B*58:20 DRB1*04:32 DRB1*08:16 DRB1*14:14 DRB1*14:20 DRB1*14:49 DPA1*02:18 DPB1*89:01 DPB1*94:01 DQA1*01:25 DQA1*04:05</errormessage>
````

Podle infa na webu a toho vystupu to pocita crossmatch proti populaci 10000 pacientu, coz je asi pro nejake rare allely hodne malo a predpokladam ze kvuli tomu dostavame ty error hlasky, ktere jsou snad jen nesikovne napsane.

Mame treba error `Warning: antigen(s) invalid and not used in the calculation: A*02:09 ...`, ale `A*02:04` to vzalo bez problemu, takze si nemyslim ze je to spatnym formatem, spis ta allela proste v populaci 10k pacientu neni. 

Muzes si ten pripad zkusit tady: https://www.etrl.org/vPRA.aspx se vstupem
```
A*02:04, A*02:09, A*02:20, A*11:09, A*24:20, A*25:01, A*66:03, A*68:08, B*14:06, B*15:146, B*15:20, B*15:46, B*35:44, B*39:01, B*44:06, B*48:04, B*48:07, B*48:10, B*58:20, DRB1*01:01, DRB1*03:01, DRB1*03:40, DRB1*04:23, DRB1*04:26, DRB1*04:32, DRB1*11:14, DRB1*11:37, DRB1*13:66, DRB1*14:14, DRB1*14:19, DPB1*92:01, DQA1*02:01, DQB1*02:02, DQB1*06:12
```

**Comment by jakubmonhart** [comment]
Tady popisuji https://etrl.eurotransplant.org/resources/hla-tables/ , ze maji tabulky allel ktere rozeznavaji (napriklad pro A https://etrl.eurotransplant.org/resources/hla-a/) .. Rychlym pohledem to vypada ze ty co nam davaji error v tabulce nejsou.

Zajimave je, ze pro DQA, DPA, DPB (https://etrl.eurotransplant.org/resources/hla-dqa1/, https://etrl.eurotransplant.org/resources/hla-dpa1/, https://etrl.eurotransplant.org/resources/hla-dpb1/) rozeznavaji HIGH RES a pak az BROAD. A BROAD zapisuji v jinem tvaru nez my , takze pokud mame pacienta ktery neni high res (nevim jak caste to je v prd), tak nam ho ted nastroj nebere.

Aby se to nedelo, jeste tam doplnim at se ty DQA, DPA, DPB davaji v broadu pokud nemame high res a dava se to v tom jejich formatu.

File: local_testing_utilities/aboi_hlai_vpra_report.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+def calculate_highly_sensitized_recipients(txm_event):
+    vpras = []
+    print('Computing vpras')
+    for recipient in tqdm(txm_event.all_recipients):
+        if recipient.db_id in txm_event.active_and_valid_recipients_dict:
+            unacceptable_antibodies = []
+            for antibodies_group in recipient.hla_antibodies.hla_antibodies_per_groups:
+                for antibody in antibodies_group.hla_antibody_list:
+                    if antibody.type == HLAAntibodyType.NORMAL and \
+                            antibody.second_code is None and \
```
Comment (jakubmonhart) [inline local_testing_utilities/aboi_hlai_vpra_report.py:104]: Tady si nejsem jisty jestli to pro potreby toho reportu filtrovat takhle, nebo tam zahrnovat i theoretical antibodies. Ten pripad kdy parsujeme na double antibody + theoretical antibodies pak totiz zpracovavame tak, ze resime jestli je tam match s tim double antibody (donor musi mit oba kody v hla typing) a theoretical match je spis upozorneni ze by tam neco mohlo byt ne?

**Comment by jakubmonhart** [comment]
Zatim takhle. Pridal jsem tam i file `vpras.csv` , at vime jak se ty hodnoty pohybuji. Tam kde je vpra==0.0 je recipient ktery ma prazdny antibodies list (to je vporadku)?

File: local_testing_utilities/aboi_hlai_vpra_report.csv [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+38-TXM-2022-07,37,19,22,4,37,23,18
+33-TXM-2021-04-v2,34,13,26,7,32,24,20
+33-TXM-2021-04,40,17,29,9,37,25,20
+39-TXM-2022-10,31,16,20,6,30,23,19
+34-TXM-2021-07,38,15,30,8,37,24,20
+35-TXM-2021-10,41,18,27,6,39,27,22
+40-TXM-2023-01,37,19,27,10,36,25,21
+41-TXM-2023-04,36,17,24,6,35,15,11
+42-TXM-2023-07,41,18,28,6,40,18,16
+all_events,494,219,351,98,472,321,250
```
Comment (kubantjan) [inline local_testing_utilities/aboi_hlai_vpra_report.csv:16]: mezi eventy jsou pacienti co jsou stejny. Tak to bych jen vypichnul u toho all_events, ze to cislo je potreba brat s rezervou

**Review by kubantjan — COMMENTED** [review]
Maly komentare. jinak ok, popprosim ted udelat ty mini upravy (zkusit to znovu pustit, pokud vse vyjde ok, tak pripravit text toho reportu)

Diky

**Comment by kubantjan** [comment]
> Zatim takhle. Pridal jsem tam i file `vpras.csv` , at vime jak se ty hodnoty pohybuji. Tam kde je vpra==0.0 je recipient ktery ma prazdny antibodies list (to je vporadku)?

@jakubmonhart ano, to je ok

File: local_testing_utilities/aboi_hlai_vpra_report.csv (hunk 2 above)
Comment (jakubmonhart) [reply local_testing_utilities/aboi_hlai_vpra_report.csv:16]: Hmm, dam ten radek vlastne spis pryc. I kdyz se opakuji pacienti, muze byt asi zajimavy prumer toho procenta incompatible paru, ale to s tech all events stejne nebude pocitat, tak budou lepsi raw pocty pres eventy a at si z toho Xenie vyjede jaka cisla chce ne?

**Comment by jakubmonhart** [comment]
**1. How many pairs of the joint pool are ABOi, HLAi, or both
ABOi and HLAi?**

**2. What is the fraction of the pairs in the joint pool that are
highly-sensitized (cPRA>90%) and very highly-sensitized (cPRA>97%)?
If you could provide average values or a table with data for each
separate matching run, it would be a great help!**

Answers to questions 1. and 2. are in the `aboi_hlai_vpra_report.csv`. There is a row for each txm event (matching run). Some of the patient pairs occur in multiple matching runs so summing the rows won't give you the absolute numbers for the joint pool (@kubantjan  jak často se to děje?). Let me know if you would be interested in that as well.

**3. Could you specify which criteria are used to identify which
exchanges will proceed for the transplantation? How do you apply
these criteria, e.g. is it sequential, or are the weights provided
for each transplants based on some characteristics (if so, which)?**

We first check sequentially if the transplant is possible. This is done based on several criteria:

Donor must have blood group that is compatible with the recipient, or a blood group that is acceptable for the recipient (acceptable blood groups are sometimes specified as the transplant can proceed even in case of ABOi)
There must be no positive crossmatch of donor's hla antigens and recipient's hla antibodies.
Exchanges between some countries are not possible.
If transplant is possible, we compute compatibility index of the donor and recipient based on how well do their HLA typings match. We sometimes further check if this compatibility index is higher then the compatibility index between the donor that is related to the recipient and do not consider the exchange otherwise.

Remaining possible exchanges are scored based on the compatibility indices (+ some bonus for blood group compatibility) and an optimal matching is found.

**Comment by kubantjan** [comment]
@jakubmonhart  deje se to fakt casto, rekl bych ze klidne 60% mezi runy zustava, obcas klidne i vic

**Comment by kubantjan** [comment]
jinak super, jdu to poslat.

**Review by kubantjan — APPROVED** [review]

**Comment by jakubmonhart** [comment]
Promin, jeste jsem z historie mazal to heslo k vpra calculatoru, at ho tu nenechavame takto verejne a ted to potrebuje znovu review .

**Review by kubantjan — APPROVED** [review]

**Pull request merged**